### ISSUE TITLE
Growable arenas

### ISSUE THREAD

**Issue by cmuratori** [4coder-editor/4coder#2]
Growable arenas would be very good, both inside the core and in the custom layer.  I think they would simplify a lot of the code everywhere, since right now there's a lot of prepassing to figure out how big things need to be, or fixed limits that need to be announced up front.  For things like the lexer, you would just pass in the arena to the lexer, and for people who don't care, you just pass a growable arena.  Now you don't have to care how big the thing is, it just works, and everything is also automatically freed together when the user frees the arena.  Listers would work this way too.  And custom layers could just ask for growable arenas for things like my project indexer.

Similarly, you had a complaint about threading and arenas, but actually growable arenas are great for threading.  The way I use them is that I assign each thread its own growable "temp" arena at startup, and then in stead of the "let me grab the global partition" code that you use, I just have my threads begin/end their temp work on their thread growable arena.  This pretty much solves all basic scratch problems.  Since entry in the custom layer already has to have an Application_Links pointer, the thread scratch arena could be on there.

Basically, I can't think of a reason why an editor _wouldn't_ want to use growable arenas.  Especially on 64-bit platforms, where you can even guarantee that the growable arenas are continuous if for some reason you really need that (like to create dynamically sized arrays that are still contiguous in memory or something).

\- Casey

**Comment by 4coder-editor** [comment]
I addressed this by building an Arena on top of the existing non-growing Partition code.  The Arena works basically the same way as the partition, but calls the low level allocator in app_links when it does not have enough memory in the underlying Partition.  I didn't change much of the core to the Arena, as there's a lot of it and it's all working well enough for now.  I did change over the lister system, which was particularly over-complicated by the Partition limit.  Other systems will change over to the Arena as I visit them.

**Comment by cmuratori** [comment]
Not sure if I know exactly what you mean, but: it is important that code doesn't need to know whether they have been handed a growing or non-growing arena.  So ideally you just want one type that is "source of memory", and all you know if that you get back what you ask for or 0.

Does that seem reasonable?

Maybe that's what you already did, though.

\- Casey

**Comment by 4coder-editor** [comment]
Is that so?  I would have thought about it the other way around "this code asks for a growing arena" vs "this code asks for a fixed size arena".  Then if you have a growable arena you can make a call that says "give me a fixed size version of this arena" which lets you call anything with the growable type, but doesn't let you call things that require growable with non-growable.  Maybe that's a bad idea anyway?  Hope it makes sense.  I can just write a wrapper type of course.

**Comment by cmuratori** [comment]
I guess I don't know what it means for "code to ask for a fixed size arena".  Code generally doesn't care what the arena is, they just want it to give them memory.  So anyone who takes a memory_arena just asks it for memory, and either they get it, or they get 0.  If the code is written to handle 0, then it could be said that it "supports fixed size arenas" in that it won't crash on out-of-memory, but then again you could also just say that it is "well behaved" because a growing arena can _also_ run out of memory eventually.

Am I missing something?

Stated alternately, it's very unusual for usage code to know or care how much memory is available, so it doesn't make a lot of sense to type the code based on that.  In very specific circumstances a piece of code might want to change its behavior based on how much memory it believes to be available, but this sort of thing can be supported (in the rare case that it happens) by just having a "how much is left" call on the arena, which returns "infinity" or the size of the VMS or whatever in the case of a growable arena.  But I've never _actually_ found I need that.

\- Casey

**Comment by 4coder-editor** [comment]
Basically I wanted to avoid having to revisit ALL of the code on the old push allocator, and I have some things which assume they are getting contiguous memory from the push allocator.  I suspect over time I can transition most or all things off it.

Maybe a reserve "a lot" up front push allocator would work better for me, but I'm a bit confused by the details of that.  In the limit it seems like that's just kicking the can down the road?

**Comment by cmuratori** [comment]
Well, I guess the question would be, does it matter if this runs on 32-bit?  Because for 64-bit, you can probably just ignore the problem entirely.  You could make every arena start out with its first allocation being 16 gigabytes away from the previous arena.  Then "allocate more" just always allocates space immediately following the last block for that arena, so it's always contiguous.

This also has the benefit of never wasting any space at the end of blocks, since you can "always use all the memory you've allocated."

\- Casey
